onionshare/web/chat_mode.py

```diff
@@ -0,0 +1,123 @@
+from flask import (
+    Request,
+    request,
+    render_template,
+    make_response,
+    flash,
+    redirect,
+    session,
+)
+from werkzeug.utils import secure_filename
+from flask_socketio import emit, join_room, leave_room
+
+
+class ChatModeWeb:
+    """
+    All of the web logic for chat mode
+    """
+
+    def __init__(self, common, web):
+        self.common = common
+        self.common.log("ChatModeWeb", "__init__")
+
+        self.web = web
+
+        # This tracks users in the room
+        self.connected_users = []
+
+        # This tracks the history id
+        self.cur_history_id = 0
+
+        self.define_routes()
+
+    def define_routes(self):
+        """
+        The web app routes for chatting
+        """
+
+        @self.web.app.route("/")
+        def index():
+            history_id = self.cur_history_id
+            self.cur_history_id += 1
+            session["name"] = (
+                session.get("name")
+                if session.get("name")
+                else self.common.build_username()
+            )
```

Comment:
This logic is a little broken and needs to be fixed. So `flask-socketIO` seems to create a fork of the actual flask session and when modified, it doesn't modify the actual session but the forked session.

>If a SocketIO handler modifies the session, the modified session will be preserved for future SocketIO handlers, but regular HTTP route handlers will not see these changes. Effectively, when a SocketIO handler modifies the session, a “fork” of the session is created exclusively for these handlers. The technical reason for this limitation is that to save the user session a cookie needs to be sent to the client, and that requires HTTP request and response, which do not exist in a SocketIO connection.

Hence if a user updates their name in chat, on refresh, their old name is shown, instead of the new name, and also causes the user list to break a bit.

There are two ways I can go about it.
1. Maintain a key: value dictionary of originally set name in the flask session as key and the updated username and any future data (like if we decide to add colors to each users) related to that user in the value. On refresh, we just reassign all those values from the dict based on the originally set name. This is a bit tricky.
2. Send an ajax request instead of a socket emit when updating the username, so that there is a HTTP Request that can modify the actual flask session variables.

I am leaning towards 2 right now.